src/mbgl/renderer/source_state.cpp

```diff
@@ -0,0 +1,157 @@
+    return result;
+}
+
+void SourceFeatureState::coalesceChanges(std::vector<RenderTile>& tiles) {
+    LayerFeatureStates changes;
+    for (const auto& layerStatesEntry : stateChanges) {
+        const auto& sourceLayer = layerStatesEntry.first;
+        FeatureStates layerStates;
+        for (const auto& featureStatesEntry : stateChanges[sourceLayer]) {
+            const auto& featureID = featureStatesEntry.first;
+            for (const auto& stateEntry : stateChanges[sourceLayer][featureID]) {
+                auto& stateKey = stateEntry.first;
+                auto& stateVal = stateEntry.second;
+
+                auto currentState = state[sourceLayer][featureID].find(stateKey);
+                if (currentState != state[sourceLayer][featureID].end()) {
+                    currentState->second = stateVal;
+                }
+                else {
+                    state[sourceLayer][featureID].insert(std::make_pair(stateKey, stateVal));
+                }
+            }
+            layerStates[featureID] = state[sourceLayer][featureID];
+        }
+        changes[sourceLayer] = layerStates;
+    }
+
+    for (const auto& layerStatesEntry : deletedStates) {
+        const auto& sourceLayer = layerStatesEntry.first;
+        FeatureStates layerStates = { {}, {} };
+
+        if (deletedStates[sourceLayer].empty()) {
+            for (const auto& featureStatesEntry : state[sourceLayer]) {
+                const auto& featureID = featureStatesEntry.first;
+                layerStates[featureID] = {};
+                state[sourceLayer][featureID] = {};
+            }
+        } else {
+            for (const auto& feature : deletedStates[sourceLayer]) {
+                const auto& featureID = feature.first;
+                bool deleteWholeFeatureState = deletedStates[sourceLayer][featureID].empty();
+                if (deleteWholeFeatureState) {
+                    state[sourceLayer][featureID] = {};
+                } else {
+                    for (const auto& stateEntry : deletedStates[sourceLayer][featureID]) {
+                        state[sourceLayer][featureID].erase(stateEntry.first);
+                    }
+                }
+                layerStates[featureID] = state[sourceLayer][featureID];
+            }
+        }
+
+        changes[sourceLayer] = layerStates;
+    }
+
+    stateChanges.clear();
+    deletedStates.clear();
+
+    if (changes.size() == 0) return;
+
+    for (auto& tile : tiles) {
+        tile.setFeatureState(changes);
+    }
+}
+
+void SourceFeatureState::removeState(const optional<std::string>& sourceLayerID, const optional<std::string>& featureID, const optional<std::string>& stateKey) {
+    std::string sourceLayer = sourceLayerID ? *sourceLayerID : "";
+
+    bool sourceLayerDeleted = deletedStates.count(sourceLayer) > 0;
+    if (sourceLayerDeleted) return;
+
+    if (stateKey && featureID) {
+        if (deletedStates[sourceLayer].count(*featureID) == 0) {
```

Comment:
this is always true, right? if there are no `sourceLayer` in deletedStates, then deletedStates[sourceLayer] will create empty entry and count() will return 0.